metric/producer.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metric
+
+import (
+	"sync"
+)
+
+// Producer is a source of metrics.
+type Producer interface {
+	// Read should return the current values of all metrics supported by this
+	// metric provider.
+	// The returned metrics should be unique for each combination of name and
+	// resource.
+	Read() []*Metric
+}
+
+// Registry maintains a set of metric producers for exporting. Most users will
+// rely on the DefaultRegistry.
+type Registry struct {
+	mu      sync.RWMutex
+	sources map[*uintptr]Producer
```

Comment:
sources map[Producer]struct{}?

Reply:
Producer may not be suitable as a map key. This way, we don't have to care.

Reply:
In what case? Also why a pointer to uintptr?